pymeasure/display/widgets/table_widget.py

```diff
@@ -0,0 +1,713 @@
+#
+# This file is part of the PyMeasure package.
+#
+# Copyright (c) 2013-2023 PyMeasure Developers
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+#
+
+import logging
+from numpy import float64, NaN
+from functools import partial
+import pyqtgraph as pg
+import pandas as pd
+
+from ..Qt import QtCore, QtWidgets, QtGui
+from .tab_widget import TabWidget
+from ...experiment import Procedure
+
+SORT_ROLE = QtCore.Qt.ItemDataRole.UserRole + 1
+SORTING_ENABLED = True  # Allow to disable sorting, for debug purpose only
+
+log = logging.getLogger(__name__)
+log.addHandler(logging.NullHandler())
+
+
+class ResultsTable(QtCore.QObject):
+    """ Class representing a panda dataframe """
+    data_changed = QtCore.Signal(int, int, int, int)
+
+    def __init__(self, results, color, column_index=None,
+                 force_reload=False, wdg=None, **kwargs):
+        super().__init__()
+        self.results = results
+        self.color = color
+        self.force_reload = force_reload
+        self.last_row_count = 0
+        self.wdg = wdg
+        self.column_index = column_index
+        self.data = self.results.data
+        self._started = False
+
+    @property
+    def data(self):
+        return self._data
+
+    @data.setter
+    def data(self, value):
+        self._data = value
+        if self.column_index is not None:
+            self._data = self._data.set_index(self.column_index)
+        else:
+            self._data.reset_index()
+
+    @property
+    def rows(self):
+        return self._data.shape[0]
+
+    @property
+    def columns(self):
+        return self._data.shape[1]
+
+    def init(self):
+        self.last_row_count = 0
+
+    def start(self):
+        self._started = True
+
+    def stop(self):
+        self._started = False
+
+    def update_data(self):
+        if not self._started:
+            return
```

Comment:
Could you explain the reason for this `_started`  logic? When looking at this in the debugger I could not find a situation (big chance that there is, but I just didn't happen upon one) where `update_data` is called while `_started=False`, or (for one case after loading data from a file) where it seems to be a problem that `update_data` is called.

Reply:
@CasperSchippers, this mechanism is used when I use `beginResetMode`/`endResentModel `calls.
I call `stop` method to stop updating the data while model is being reset.
Similarly, I call `start` method to start data update after new `ResultsTable` has been added to the results list.

This is to avoid inconsistency while model is being updated.

Reply:
Ok, thanks; I'm fine with leaving it as it is to ensure no inconsistencies occur.